proposals/0003-external-calls.md

```diff
@@ -0,0 +1,344 @@
+
+Educated calls meanwhile are not guaranteed to not introduce errors, but they have certain guarantees attached. These means that it combats (1), (2), (4):
+1. That there is a defined function at the end of the call
+2. That function obeys the modifiers given
+3. That function has the same return types and parameter types as defined
+4. The contract you call matches the source code given
+5. The gas provided should be inferred by gas estimation over flint
+
+We propose a method to both declare this interface within Flint, use the The Flint Package Manager to extract an interface, or call contracts distrusted.
+
+In _Distrusted Calls_, the contract (Client), sets the properties of the director which then sets up the command which is finally sent to the contract.
+The aim is to encapsulate a request as an object, thereby letting Flint parametrize clients with different requests.
+
+(3) is combated by having necessary catching of all external calls - or prefixing with `try!` and `try?` to revert and nullify respectively.
+
+### Distrusted Calls
+#### Interface specified
+```swift
+interface Alpha(State1, State2) {
+  var owner: Address
+
+  Alpha @(State1) :: (any) {
+    func doesNothing()
+  }
+  Alpha @(any) :: (owner) {
+    func doesNothingWithArgs(Int, Int, Int)
+    func withdraw() -> Int
+    func deposit(Int) -> Bool
+  }
+  @payable
+  func expensiveFunction()
+}
+
+let alpha: Director<Alpha> = 0x000... with Alpha
+
+try alpha!.doesNothing() {
+  // Successful Call
+}
+catch {
+  // If it fails
+}
+try! alpha!.doesNothingWithArgs(x, y, z)
+// If catch can not be provided if try! is used and then the transaction reverts on failure
+
+try! alpha!.withdraw() // This flags an error as the return value is not dealt with
+
+try! boundReturn <- alpha!.getReturn() {
+  // Optionally does something with boundReturn
+}
+
+// Setting contract instance properties
+alpha!.value = Wei(200)
+alpha!.gas = Gas(2000)
+alpha.trust() // Removes the need for a bang
+try! alpha.expensiveFunction()
+```
+#### Foreign Function Interface
+Flint smart contract can call functions from Solidity smart contracts and vice-versa, thanks to the Flint Foreign Function Interface (FFI).
+
+The Flint FFI allows smart contracts to import a Solidity contract, in order to statically check the validity of external function calls.
+
+```swift
+@foreign import ForeignContract // A Solidity contract
+
+contract Foo {}
+
+Foo :: (any) {
+  func foo(address: Address) -> Int {
+    return (address as ForeignContract)!.getValue()
+  }
+}
+```
+
+
+### Trusted Calls
+
+#### Flint Package Manager
+```swift
+// Creates a contract from the data stored in Flint Package Manager
+var tokenInstance: Contract<ERC.Token> = Flint.new(0x000...)
```

Comment:
It feels like it would be better to extend the below for FPM, rather than something completely different.